Grain v0.4, Cereal, blog post (#250)
grain-lang/grain-lang.org: src_blog/_posts/New-Release-Grain-v0.4.0-Cereal.md

---

> title: New Release! Grain v0.4.0 - Cereal

Reviewer: Maybe we should leave it as `v0.4`, since the _real_ release is `0.4.2`?

Author: ```suggestion
title: New Release! Grain v0.4 - Cereal
```

> subtitle: Announcing the release of Grain v0.4.0, Cereal, with NEAR support.

Author: ```suggestion
subtitle: Announcing the release of Grain v0.4, Cereal, with NEAR support.
```

date: 2021-09-07 12:00:00
author: Blaine Bublitz

> cover: /blog/photos/cover/marie-bellando-mitjans-cBChXxQqEKM-unsplash.jpg

Reviewer: This was like the first result I found, but I actually like it https://unsplash.com/photos/EV6C1LjH1Lk

Reviewer 2: Lots of Grain yellow, I like it too

Author: Used that one! Thanks!

coverAttribution: Marie Bellando-Mitjans / Unsplash
tags: Release, Changelog
---

> After nearly 5 months of hard work, we are excited to finally share Grain v0.4.0 with the world! This release marks the culmination of a bunch of parallel features being built by the core team and a ton of amazing contributions from our community. 🎉

Reviewer: ```suggestion
After nearly 5 months of hard work, we are excited to finally share Grain v0.4.0 with the world! This release marks the culmination of a bunch of exciting features being built by the core team and a ton of amazing contributions from our community. 🎉
```
style nit (didn't feel like "parallel" was a good word here)

Author: ```suggestion
After nearly 5 months of hard work, we are happy to finally share Grain v0.4 with the world! This release marks the culmination of a bunch of exciting features being built by the core team and a ton of amazing contributions from our community. 🎉
```

Author: didn't want to say exciting and excited in the same paragraph

> As is typical with these major announcements, I will share some of the highlights we are excited about, but you should check out all of the awesome in the changelog attached to our [tagged GitHub release](https://github.com/grain-lang/grain/releases/tag/grain-v0.4.0).

Reviewer: ```suggestion
As is typical with these major announcements, I will share some of the highlights we are particularly excited about, but you should check out all of the awesome updates in the changelog attached to our [tagged GitHub release](https://github.com/grain-lang/grain/releases/tag/grain-v0.4.0).
```

## NEAR Support

> A big portion of the work that went into this release was preparing the compiler to output WebAssembly that will run on the NEAR blockchain. NEAR already supports Rust (compiled to wasm) and AssemblyScript, but any language that targets WebAssembly should be able to output smart contracts that will run on their platform. We needed to make some adjustments to the compiler itself to support intricacies with the platform, but the NEAR Foundation graciously provided us with [a $50,000 grant](https://grain-lang.org/blog/2021/08/16/grain-awarded-50000-grant-from-near-foundation/) to make this work possible!

Reviewer: ```suggestion
A big portion of the work that went into this release was preparing the compiler to output WebAssembly that is compatible with the NEAR blockchain. NEAR already supports Rust (compiled to Wasm) and AssemblyScript, but any language that targets WebAssembly should be able to output smart contracts that can run on their platform. We needed to make some adjustments to the compiler itself to support the intricacies of the platform, but the NEAR Foundation graciously provided us with [a $50,000 grant](https://grain-lang.org/blog/2021/08/16/grain-awarded-50000-grant-from-near-foundation/) to make this work possible!
```

> We have preliminary support for NEAR in this release, but things will get **really** interesting once we write the NEAR SDK for Grain. This is some of our most pressing work now that the release is out—check back soon to start writing smart contracts in Grain. 📈

Reviewer: ```suggestion
We have preliminary support for NEAR in this release, but things will get **really** interesting once we release the NEAR SDK for Grain. This is some of our most pressing work now that the release is out—check back soon to start writing smart contracts in Grain. 📈
```

## Tools

> Another important goal for this release was to improve the tooling experience for our end users and for ourselves. Towards this goal, we added two new commands to the Grain CLI: `grain format` and `grain doc`.

Reviewer: ```suggestion
Another important goal for this release was to improve the tooling experience for both our end users and ourselves. To this end, we have added two new commands to the Grain CLI: `grain format` and `grain doc`.
```

> ### Grain format

Reviewer: ```suggestion
### Grainformat
```
Consistent w/ blog post

> Grain format provides an opinionated code formatter tool for Grain source code. This tool was an absolutely herculean effort by Marcus, the core team member we introduced in our last release post and maintainer of the Grain Language Server Protocol.

Reviewer: ```suggestion
Grainformat provides an opinionated code formatter tool for Grain source code. This tool was an absolutely herculean effort by [Marcus](https://github.com/marcusroberts), the core team member we introduced in our last release post and maintainer of the Grain Language Server Protocol.
```
🤷 on the GH profile link; might be nice for us to do for shout-outs

Author: I trend towards twitter
```suggestion
Grain format provides an opinionated code formatter tool for Grain source code. This tool was an absolutely herculean effort by [Marcus](https://twitter.com/marcusr), the core team member we introduced in our last release post and maintainer of the Grain Language Server Protocol.
```

> I won’t go in-depth on the formatter in this post, but Marcus wrote up an excellent [blog post](https://grain-lang.org/blog/2021/09/04/grain-formatter/) about it! I recommend you give it a read, and grab our latest vscode extension to have format-on-save; I’ve been using it for a few days and it is a great experience.

Reviewer: ```suggestion
I won’t go in-depth on the formatter in this post, but Marcus wrote up an excellent [blog post](https://grain-lang.org/blog/2021/09/04/grain-formatter/) about it! I recommend you give it a read, and grab our latest VSCode extension to get access to format-on-save; I’ve been using it for a few days and it is a great experience.
```

> ### Grain doc

Author: ```suggestion
### Graindoc
```

> Additionally, I finally finished the initial work around our documentation tool. This provides a JSDoc-like experience for your Grain code. We’ve already begun using it to document the standard library, so we’ll be able to keep our API docs consistently updated.

Reviewer: ```suggestion
Additionally, I finally finished the initial work around our documentation tool. This provides a [JSDoc](https://jsdoc.app/)-like experience for your Grain code. We’ve already begun using it to document the standard library, so we’ll be able to keep our API documentation consistently updated.
```

> You can check out the features and syntax in my preview [blog post](https://grain-lang.org/blog/2021/08/27/tooling-preview-graindoc/). I’m particularly happy that the tool infers the types of your parameters and return values, so you don’t have to worry about keeping them in-sync.

Reviewer: ```suggestion
You can check out the features and syntax in my preview [blog post](https://grain-lang.org/blog/2021/08/27/tooling-preview-graindoc/). I’m particularly happy that the tool infers the types of your parameters and return values, so you don’t have to worry about keeping them in-sync with updates to your APIs.
```

## Callee-owned values

> This is a bit of an internal change, but it can cause a lot of ripples throughout the ecosystem. We changed the way our compiler outputs Garbage Collector code so that values passed to a function are owned and cleaned up by the function being called. Previously, the caller would `decRef` the values after a function returned; however, we found that this caused some memory leaks in certain scenarios. By changing this memory ownership convention, we now know all parameters are `decRef`ed by the time the function returns.
> 
> This only has ramifications to you if you were using the `@disableGC` annotation on your Grain code. While this attribute is not recommended for end-user code, we use it in the stdlib and some library authors might need to use it. If you are in need of this, please get in touch with us to understand the new calling convention.

Reviewer 3: I think we should go into a lot less detail here. It's worth calling out, but I don't think we should really get into it in this post. I think a technical post about it (hi @peblair) would be good.
```suggestion
This is a bit of an internal change. We changed Grain's garbage collection calling convention such that a function that is called owns the references to its arguments rather than the caller. This was done to help us assure that all allocated memory is properly reclaimed.

This shouldn't at all affect how you use Grain! This only has ramifications to you if you were using Grain's `unsafe` libraries for low-level code or bindings to WebAssembly host functions. This is typically only done by some library authors (and isn't recommended otherwise). If you are actively using the `unsafe` libraries, please get in touch with us to understand the new calling convention.
```

## Standard library

As with most releases, we had a lot of work happening in the standard library—some by the core team and some by community contributors!

### Number

> Thanks to [@spotandjake](https://twitter.com/spotandjake), we finally have a preliminary module for working with our Number type. This module will probably be one of the most used, so we are glad for the contribution. You can find the docs in our [standard library docs](https://grain-lang.org/docs/stdlib/number).

Reviewer: ```suggestion
Thanks to [@spotandjake](https://twitter.com/spotandjake), we finally have a dedicated module for working with our Number type. This module will probably be one of the most used, so we are glad for the contribution. You can find the docs in our [standard library documentation](https://grain-lang.org/docs/stdlib/number).
```

### Sys

> Previous to this release, any WASI calls in our `sys/*` modules would throw if they didn’t return a success code. This was a problem for end users because you couldn’t catch the error and handle it, such as when trying to `stat` a file to see if it exists. We’ve changed all of these functions to return a `Result` type instead, so you can react to the errors.

Reviewer: ```suggestion
Prior to this release, any WASI calls in our `sys/*` modules would throw an error if they didn’t return a success code. This was a problem for end users because you couldn’t catch the error and handle it, such as when trying to `stat` a file to see if it exists. We’ve changed all of these functions to return a `Result` type instead, so you can decide how to deal with these errors.
```

Reviewer 3: Combined @peblair's suggestion into mine:
```suggestion
Prior to this release, any system calls via WASI, the WebAssembly System Interface, in our `sys/*` modules would throw an error if they didn’t return a success code. As there isn't yet support for WebAssembly exceptions, this wasn't great because you couldn’t catch the error and handle it, such as when trying to `stat` a file to see if it exists. We’ve changed all of these functions to return a `Result` type instead, so you can now handle any errors that occur.
```

### Buffer

> An amazing effort by [@jozanza](https://twitter.com/jozanza), the Buffer module provides a data type for storing bytes that grows as data is appended. This can be used for efficiently concatenating large amounts of data, such as when building JSON strings, instead of allocating excessive amounts of strings to combine.

Reviewer: ```suggestion
An amazing effort by [@jozanza](https://twitter.com/jozanza), the Buffer module provides a new data type for storing bytes which automatically grows as data is appended. This can be used for efficiently concatenating large amounts of data, such as when building JSON strings, instead of allocating excessive amounts of strings to combine.
```

## Welcoming Josiah

> I’m also excited to announce that we’ve added [Josiah](https://twitter.com/jozanza) to the Core Team! Oscar, Philip, Marcus, and I are extremely excited to have him join the team. In addition to the Buffer module, Josiah has also contributed our Bytes module, contributed to our Binaryen.ml bindings, and has some really cool Grain-related projects in the works. Welcome to the team, Josiah! 🙌

Reviewer: ```suggestion
I’m also excited to announce that we’ve added [Josiah](https://twitter.com/jozanza) to the Core Team! Oscar, Philip, Marcus, and I are extremely excited to have him join the team. In addition to the Buffer module, Josiah has also contributed our Bytes module, contributed to our [Binaryen.ml bindings](https://github.com/grain-lang/binaryen.ml), and has some really cool Grain-related projects in the works. Welcome to the team, Josiah! 🙌
```
gotta bump that SEO
